### DISCUSSION TITLE
Can a snap get the wallet address and also sign some text?

### DISCUSSION THREAD

**Discussion by javaspeak** [MetaMask/snaps#1848] [Q&A]
Can a snap:

(i)  get the wallet address
(ii) sign some text
(iii) send that signed text and address to an external api to login to that api?

If yes where do I find the API for that - currently cannot find that functionality in these docs:

     https://docs.metamask.io/snaps/

My work around would be to get the dapp to get the address and sign the text and then pass the address and signature to the snap with a rpc call.  The snap could then pass this info to the api so that it can get some trust info on the "to" address of the transaction.

The idea is this snap intercepts a transaction and warns the user if some people of the user's network have reported an issue for the contract "to" address in the past.

Any help / advice much appreciated and thank you for your time.

**Comment by Montoya** [comment]
Hi @javaspeak , if you have the permission `endowment:ethereum-provider` then you can do `eth_requestAccounts` from within your Snap, then with the connected account(s) you can do `personal_sign`. It works very similar to how the dapp API works for a dapp. 

https://docs.metamask.io/snaps/reference/permissions/#dynamic-permissions

The only thing is you can't intercept transactions right now and you can't request accounts from a transaction insight entrypoint.

**Reply by javaspeak** [reply]
Thankyou @Montoya 

If you cannot intercept transactions then what is this function for?:

```
export const onTransaction: OnTransactionHandler = async ({ transaction }) => {

}
```

Is this onTransactionHandler called after the transaction?

To me the naming start with "On" suggests that it is called during the transaction instead of after?

If it is called after the transaction then we need a new SIP to be able to intercept the transaction before it is closed off.

Is there something am missing here?

Thank you

**Reply by javaspeak** [reply]
Also what is a "transaction insight entrypoint?"

**Reply by Montoya** [reply]
onTransaction is called by MetaMask to a transaction insight Snap. To see how it works, you can follow this tutorial: https://docs.metamask.io/snaps/tutorials/transaction-insights/

You can also try a transaction insight Snap, like Wallet Guard: https://snaps.metamask.io/snap/npm/wallet-guard-snap/

> To me the naming start with "On" suggests that it is called during the transaction instead of after?

That's correct! "Transaction insight entrypoint" is one of the Snaps entrypoints. An entry point is where a Snap can be invoked to run and either provide a response or do some sort of execution. Snaps do not run endlessly, they have to be invoked by something to run and that is an entrypoint. "Transaction insight" is a specific entrypoint where MetaMask can call a Snap to give information about a transaction. 

If you want to build a Snap that can just warn the user if the "to" of a transaction send or contract interaction is malicious, you can already do that. You can follow the tutorial I provided above. Here's an example Snap you can try in Flask that uses the "to" parameter: https://github.com/Montoya/insights-counter
